### PR TITLE
 Fix insertion to inflight_dma_nodes

### PR DESCRIPTION
**Pull request by Sacusa**
When the queue `inflight_dma_nodes` is full and more than `maxInflightNodes` new DMA nodes are executed, the new requests will wait indefinitely since the condition `inflight_dma_nodes.size() < maxInflightNodes` will never evaluate to true.

This commit changes the logic so that new nodes are only queued when `inflight_dma_nodes` has space.

### CODE REVIEW

File: gem5/HybridDatapath.cpp
Comment (xyzsam) [inline gem5/HybridDatapath.cpp:520]: I don't think the statement "inflight_dma_nodes.size() < maxInflightNodes will never evaluate to true" is correct. Other DMA nodes could complete and erase their entries from inflight_dma_nodes. 

I'm also not sure what this PR actually changes in terms of behavior:
* Before, `handleDmaMemoryOp` would add the entry into the map and return false if the map's size > maxInflightNodes. Subsequent calls to this function will also return false since `status = Invalid` and the map's size is still too large.
* After, `handleDmaMemoryOp` won't add the entry at all; it will just return false; subsequent calls will do the same thing.

So at the end of the day, nothing has changed...am I missing something?

Comment (Sacusa) [reply gem5/HybridDatapath.cpp:520]: Consider this example:

Let us say the queue size is 100 and it is currently full right now. The system receives, say, 101 new DMA requests which are set to `Invalid` and added to the queue (making the queue size 201), but not issued since `inflight_dma_nodes.size() > maxInflightNodes`. After some time, the initial 100 requests will all complete and leave the queue, but the condition `inflight_dma_nodes.size() < maxInflightNodes` will still be false, preventing any of the new 101 requests from being issued.

You are right in that the PR doesn't change anything behaviorally, but it prevents the problem I just described by only adding nodes to the queue when they can be processed.

Comment (xyzsam) [reply gem5/HybridDatapath.cpp:520]: Ah, I see. Thanks for the explanation! Can you include a shortened version of this in a comment in the code for future readers?

Comment (Sacusa) [reply gem5/HybridDatapath.cpp:529]: Done!

**Review by xyzsam — APPROVED** [review]

**Comment by Sacusa** [comment]
I suppose this can be merged now?

**Comment by yaoyuannnn** [comment]
> I suppose this can be merged now?

Go ahead and merge. We always leave it to the author to merge after getting approval :)

**Pull request merged**
